elixir-fundamentals-review.livemd

```diff
@@ -0,0 +1,339 @@
+```
+
+## Modules and Structs
+
+In the cell below, define a `Post` struct with `:title` and `:content` keys. Enforce that these keys must be provided a value. You do not need to validate the values.
+
+```mermaid
+erDiagram
+    Post {
+        string title
+        string content
+    }
+```
+
+```elixir
+post = %{title: "book", content: "inside"}
+post
+```
+
+Create an instance of the `Post` struct you defined above.
+
+```elixir
+post
+```
+
+## Enumeration
+
+Complete the `Enumeration` module as documented.
+
+```elixir
+defmodule Enumeration do
+  @doc """
+  Double every integer in a list or range.
+
+  iex> Enumeration.double(1..5)
+  [2, 4, 6, 8, 10]
+  """
+  def double(list) do
+    Enum.map(list, fn each -> each * 2 end)
+  end
+
+  @doc """
+  Return only the even numbers in a list or range.
+
+  iex> Enumeration.evens([2, 2, 5, 4])
+  [2, 2, 4]
+  """
+  def evens(list) do
+    Enum.filter(list, fn each -> rem(each, 2) == 0 end)
+  end
+
+  @doc """
+  Sum only the even numbers in a list or range.
+
+  iex> Enumeration.sum_evens([2, 2, 5, 4])
+  8
+  """
+  def sum_evens(list) do
+    evens(list)
+    |> Enum.sum()
+  end
+end
+
+Enumeration.double(1..5)
+Enumeration.evens([2, 2, 5, 4])
+Enumeration.sum_evens([2, 2, 5, 4])
+```
+
+## Built In Modules
+
+You're going to use the built-in modules to perform the common CRUD (create, read, update, destroy/delete) operations on each data structure.
+
+<!-- livebook:{"break_markdown":true} -->
+
+### Maps
+
+<!-- livebook:{"break_markdown":true} -->
+
+Update (increment) the `:count` key in this map to return `%{count: 2}`.
+
+```elixir
+old_map = %{count: 1}
+Map.update(old_map, :count, 0, fn each -> each + 1 end)
+```
+
+Set the `:greeting` key in this map to be `"hi"` instead of `"hello"`.
+
+```elixir
+old_map = %{greeting: "hello"}
+Map.put(old_map, :greeting, "hi")
+```
+
+Delete the `:complete` key from this map to return an empty map.
+
+```elixir
+old_map = %{complete: "task"}
+Map.delete(old_map, :complete)
+```
+
+### List
+
+<!-- livebook:{"break_markdown":true} -->
+
+Delete the `"a"` element in this list.
+
+```elixir
+List.delete(["a", "b", "c"], "a")
+```
+
+Insert a `"b"` element at index `1` in this list.
+
+```elixir
+List.insert_at(["a", "c"], 1, "b")
+```
+
+Update the element at index `1` in this list to be lowercase.
+
+```elixir
+List.update_at(["a", "B", "c"], 1, fn each -> String.downcase(each) end)
+```
+
+### Keyword List
+
+<!-- livebook:{"break_markdown":true} -->
+
+Delete the `:a` key/value pair in this keyword list.
+
+```elixir
+Keyword.delete([a: 1], :a)
+```
+
+Insert a `:b` key with the value `2` in this keyword list at index `1` to create `[a: 1, b: 2, c: 3]`. Order does matter.
+
+<details style="background-color: burlywood; padding: 1rem; margin: 1rem 0;">
+<summary>Hint</summary>
+
+Remember that you can use `List` functions such as `List.insert_at/3` with keyword lists.
+
+</details>
+
+```elixir
+List.insert_at([a: 1, c: 3], 1, b: 2)
```

Comment:
This won't produce the correct result.

The following is equivalent.

```elixir
List.insert_at([a: 1, c: 3], 1, b: 2)
List.insert_at([a: 1, c: 3], 1, [b: 2])
```

To insert an element into a keyword list, you need to insert the two-element tuple `{:b, 2}` like so.

```elixir
List.insert_at([a: 1, c: 3], 1, {:b, 2})
```